File: test/test-matrix_rw.cpp
#include "matrix_rw.hpp"

using Uint_T = matrix_rw::Uint_T;
using Real_T = matrix_rw::Real_T;

//* setup
const std::string dat_dir = "../dat";
const std::string ref_dat_dir = "../../test/reference_dat";
const std::string test_name = "test-matrix_rw";
const std::string dat_prefix = dat_dir + "/" + test_name + "-";
const std::string ref_dat_prefix = ref_dat_dir + "/" + test_name + "-";
const std::string in_arr_fname = "in_arr.dat";
const std::string out_arr_fname = "out_arr.dat";

constexpr Uint_T n_dim = 150;
constexpr Uint_T m_dim = 75;

#ifdef __USE_SINGLE_PRECISION__
constexpr Real_T error_thres = 1e-7;
#else
constexpr Real_T error_thres = 1e-15;
#endif

int
main()
{
	//* read reference data
	Real_T in_arr[n_dim * m_dim];
	matrix_rw::read<n_dim, m_dim>(ref_dat_prefix + in_arr_fname, in_arr);

	//* test
	matrix_rw::write<n_dim, m_dim>(dat_prefix + out_arr_fname, in_arr);

	Real_T out_arr[n_dim * m_dim];
	matrix_rw::read<n_dim, m_dim>(dat_prefix + out_arr_fname, out_arr);

	//* verify
	Real_T max_error = 0.;

	for (Uint_T i = 0; i < n_dim * m_dim; ++i) {
		const Real_T error = std::abs(in_arr[i] - out_arr[i]);

		if (error > max_error) {
			max_error = error;
		}
	}

	if (max_error < error_thres) {
		return 0;
	} else {
		return 1;
	}
}

File: test/matlab/test__matrix_rw.m
%********
%* setup
%********
io_config;
test_name = 'test-matrix_rw';
dat_prefix = append(dat_dir, '/', test_name, '-');
exe_name = append(test_name, '.exe');
in_arr_fname = 'in_arr.dat';
out_arr_fname = 'out_arr.dat';

n_dim = 150;
m_dim = 75;

error_thres = 1e-16;

%* __USE_SINGLE_PRECISION__ 
%* error_thres = 1e-6; 

%*******************
%* create test data 
%*******************
in_arr = [magic(m_dim); magic(m_dim)]./1e3 - 1/3;

%************************************
%* write input (for test executable)
%************************************
writematrix(in_arr, append(dat_prefix, in_arr_fname), 'Delimiter', delimiter);  

%***************************
%* call the test executable
%***************************
prev_pwd = pwd;
cd(bin_dir);

if ~isfile(exe_name)
	error(append(bin_dir, '/', exe_name, ' does not exist. Use CMake to build the test.'));
end

if system(exe_name) > 0
	warning(append(bin_dir, '/', exe_name, ' has returned failure.'));
end

cd(prev_pwd);

%******************************************
%* read output (created by the executable)
%******************************************
out_arr = readmatrix(append(dat_prefix, out_arr_fname));

%*********
%* verify
%*********
max_error = max(in_arr - out_arr, [], 'all');

if max_error < error_thres
    disp(append(test_name, '	ok'));
else
    disp(append(test_name, '	fail'));
end